# Data Experimentation Capstone

## Introduction

The objective of this project is to examine the effects of user engagement after seeing Ad A or Ad B

## Problem

An advertising company has developed a new ad to have users engage with their questionnaire. The company has shown the new ad to some users and a dummy ad to others and wants their data analyst team to interpret the results. Does the new ad generate more responses to their questionnaire? Is it statistically significant? Is the company justified in using the new ad? 

A/B testing is common in the business world and is a way to compare two versions of something to figure out which performs better. Figuring out which ad users prefer is a real life business problem that would be expected to know how to solve as a business data analyst. 


## Data

Dataset is found on kaggle from an advertising company. https://www.kaggle.com/osuolaleemmanuel/ad-ab-testing

## Importing Libraries

In [1]:
import pandas as pd
import numpy as np

## Reading The Data

In [2]:
# setting dataframe
df = pd.read_csv("AdSmartABdata.csv")

In [3]:
# checking head
df.head()

auction_id experiment        date  hour  \
0  0008ef63-77a7-448b-bd1e-075f42c55e39    exposed  2020-07-10     8   
1  000eabc5-17ce-4137-8efe-44734d914446    exposed  2020-07-07    10   
2  0016d14a-ae18-4a02-a204-6ba53b52f2ed    exposed  2020-07-05     2   
3  00187412-2932-4542-a8ef-3633901c98d9    control  2020-07-03    15   
4  001a7785-d3fe-4e11-a344-c8735acacc2c    control  2020-07-03    15   

          device_make  platform_os                browser  yes  no  
0  Generic Smartphone            6          Chrome Mobile    0   0  
1  Generic Smartphone            6          Chrome Mobile    0   0  
2               E5823            6  Chrome Mobile WebView    0   1  
3   Samsung SM-A705FN            6               Facebook    0   0  
4  Generic Smartphone            6          Chrome Mobile    0   0

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8077 entries, 0 to 8076
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   auction_id   8077 non-null   object
 1   experiment   8077 non-null   object
 2   date         8077 non-null   object
 3   hour         8077 non-null   int64 
 4   device_make  8077 non-null   object
 5   platform_os  8077 non-null   int64 
 6   browser      8077 non-null   object
 7   yes          8077 non-null   int64 
 8   no           8077 non-null   int64 
dtypes: int64(4), object(5)
memory usage: 568.0+ KB


There are 8077 rows and 9 columns in the dataset.

## The Data Wrangling

The dataset contains the yes and no columns. A 1 in one the columns indicates it was selected, and a 0 indicates it wasn't.
A 0 in both indicates the user ignored the questionnaire.
Users who answered and whether they answered yes or no is most important therefore all rows with 0 in both yes/no columns will be dropped

In [5]:
drop = df.loc[(df["yes"]==0) & (df["no"]==0)]
drop.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6834 entries, 0 to 8076
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   auction_id   6834 non-null   object
 1   experiment   6834 non-null   object
 2   date         6834 non-null   object
 3   hour         6834 non-null   int64 
 4   device_make  6834 non-null   object
 5   platform_os  6834 non-null   int64 
 6   browser      6834 non-null   object
 7   yes          6834 non-null   int64 
 8   no           6834 non-null   int64 
dtypes: int64(4), object(5)
memory usage: 533.9+ KB


6834 users in the dataset did not answer the questioner so they will be dropped.

In [6]:
df.drop(drop.index, axis=0,inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1243 entries, 2 to 8071
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   auction_id   1243 non-null   object
 1   experiment   1243 non-null   object
 2   date         1243 non-null   object
 3   hour         1243 non-null   int64 
 4   device_make  1243 non-null   object
 5   platform_os  1243 non-null   int64 
 6   browser      1243 non-null   object
 7   yes          1243 non-null   int64 
 8   no           1243 non-null   int64 
dtypes: int64(4), object(5)
memory usage: 97.1+ KB


After dropping users who did not answer we have 1243 enteries left.

We will now create a new column "answer" where 0 indicates no and 1 indicates yes then drop the yes & no columns

In [7]:
# creating answer column
df["answer"] = df["yes"]

# dropping yes and no columns
df.drop(["yes","no"], axis=1, inplace=True)

# checking dataframe
df.head()

auction_id experiment        date  hour  \
2   0016d14a-ae18-4a02-a204-6ba53b52f2ed    exposed  2020-07-05     2   
16  008aafdf-deef-4482-8fec-d98e3da054da    exposed  2020-07-04    16   
20  00a1384a-5118-4d1b-925b-6cdada50318d    exposed  2020-07-06     8   
23  00b6fadb-10bd-49e3-a778-290da82f7a8d    control  2020-07-08     4   
27  00ebf4a8-060f-4b99-93ac-c62724399483    control  2020-07-03    15   

           device_make  platform_os                browser  answer  
2                E5823            6  Chrome Mobile WebView       0  
16  Generic Smartphone            6          Chrome Mobile       1  
20  Generic Smartphone            6          Chrome Mobile       0  
23    Samsung SM-A202F            6               Facebook       1  
27  Generic Smartphone            6          Chrome Mobile       0

In [8]:
# checking datatypes
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1243 entries, 2 to 8071
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   auction_id   1243 non-null   object
 1   experiment   1243 non-null   object
 2   date         1243 non-null   object
 3   hour         1243 non-null   int64 
 4   device_make  1243 non-null   object
 5   platform_os  1243 non-null   int64 
 6   browser      1243 non-null   object
 7   answer       1243 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 87.4+ KB


## Exploratory Data Analysis

In [9]:
# converting date column to datetime
df['date'] = pd.to_datetime(df['date'], infer_datetime_format=True)

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1243 entries, 2 to 8071
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   auction_id   1243 non-null   object        
 1   experiment   1243 non-null   object        
 2   date         1243 non-null   datetime64[ns]
 3   hour         1243 non-null   int64         
 4   device_make  1243 non-null   object        
 5   platform_os  1243 non-null   int64         
 6   browser      1243 non-null   object        
 7   answer       1243 non-null   int64         
dtypes: datetime64[ns](1), int64(3), object(4)
memory usage: 87.4+ KB


In [11]:
# checking count of unique values in experiment column
df.groupby('experiment')[['auction_id']].count()

auction_id
experiment            
control            586
exposed            657

There are 586 users in the control group and 657 users in the exposed group.

In [12]:
# checking count of unique values in date column
df_date=df.groupby('date')[['auction_id']].count().sort_values(by='auction_id', ascending=False)
df_date

auction_id
date                  
2020-07-03         325
2020-07-09         184
2020-07-08         177
2020-07-04         159
2020-07-10         124
2020-07-05         117
2020-07-07          83
2020-07-06          74

Most replies were made on July 3rd

In [13]:
# checking count of unique values in hour column
df_hour = df.groupby('hour')[['auction_id']].count().sort_values(by='auction_id', ascending=False)
df_hour

auction_id
hour            
15           281
8             67
7             59
10            57
9             57
14            51
6             50
5             49
4             46
20            45
11            44
3             44
16            44
12            42
13            41
18            38
19            38
21            34
17            34
2             33
1             32
0             31
22            21
23             5

The users are most active from early morning hours to afternoon hours (6-11 to 12-16).
And least active at night (19-2).

In [14]:
# checking count of unique values in devive maker column
df_device = df.groupby('device_make')[['auction_id']].count().sort_values(by='auction_id', ascending=False)
df_device

auction_id
device_make                   
Generic Smartphone         719
Samsung SM-G960F            45
Samsung SM-G950F            35
Samsung SM-G973F            22
Samsung SM-A202F            18
...                        ...
Samsung SM-G925F             1
Samsung SM-G965U1            1
Samsung SM-J330F             1
Samsung SM-J330G             1
Samsung SM-A750GN            1

[115 rows x 1 columns]

Most of the users had a Generic Smartphone or a Samsung phone.

In [15]:
# checking count of unique values in platform_os column
df_os=df.groupby('platform_os')[['auction_id']].count().sort_values(by='auction_id', ascending=False)
df_os

auction_id
platform_os            
6                  1226
5                    17

Majority of users were on Platform Os 6

In [16]:
# checking count of unique values in browser column
df_browser=df.groupby('browser')[['auction_id']].count().sort_values(by='auction_id', ascending=False)
df_browser

auction_id
browser                               
Chrome Mobile                      695
Chrome Mobile WebView              227
Facebook                           156
Samsung Internet                   145
Mobile Safari                       14
Mobile Safari UI/WKWebView           3
Chrome                               2
Chrome Mobile iOS                    1

The most used browsers when ansering the questionnaire were Chrome Mobile, Chrome Mobile WebView, Facebook and Samsung Internet.

In [17]:
# checking count of unique values in answer column
df_answer=df.groupby('answer')[['auction_id']].count().sort_values(by='auction_id', ascending=False)
df_answer

auction_id
answer            
0              671
1              572

There was overall more No (0) than Yes(1) in the replies.

## Statistical Analysis

### Experiment Approach¶
- Null Hypothesis 𝐻ₒ: p = pₒ "There is no significant difference between the ad success rate of both groups"


- Alternative Hypothesis Hₐ: p ≠ pₒ "There is significant difference between the ad success rate of both groups"


- Confidence Level: 95% (α=0.05)


- Given we don’t know if the new design will perform better/worse/equal as our current design, we will perform a two-tailed test
    

In [18]:
# selecting a sample of the dataset for equal comparison
required_n = 586
control_sample = df[df['experiment'] == 'control'].sample(n=required_n, random_state=22)
exposed_sample = df[df['experiment'] == 'exposed'].sample(n=required_n, random_state=22)

ab_test = pd.concat([control_sample, exposed_sample], axis=0)
ab_test.reset_index(drop=True, inplace=True)

## Comparing statistics of groups

In [23]:
import scipy.stats as stats
conversion_rates = ab_test.groupby('experiment')['answer']

std_p = lambda x: np.std(x, ddof=0)              # Std. deviation of the proportion
se_p = lambda x: stats.sem(x, ddof=0)            # Std. error of the proportion (std / sqrt(n))

conversion_rates = conversion_rates.agg([np.mean, std_p, se_p])
conversion_rates.columns = ['conversion_rate', 'std_deviation', 'std_error']


conversion_rates.style.format('{:.3f}')


Judging by the stats above, it does look like our exposed group design performed similarly, with our new design performing slightly better, at 45.1% vs. 45.7% conversion rate.

## Hypothesis Testing

In [24]:
from statsmodels.stats.proportion import proportions_ztest, proportion_confint

In [25]:
control_results = ab_test[ab_test['experiment'] == 'control']['answer']
exposed_results = ab_test[ab_test['experiment'] == 'exposed']['answer']

In [26]:
n_con = control_results.count()
n_treat = exposed_results.count()
successes = [control_results.sum(), exposed_results.sum()]
nobs = [n_con, n_treat]

z_stat, pval = proportions_ztest(successes, nobs=nobs)
(lower_con, lower_treat), (upper_con, upper_treat) = proportion_confint(successes, nobs=nobs, alpha=0.05)

print(f'z statistic: {z_stat:.2f}')
print(f'p-value: {pval:.3f}')
print(f'ci 95% for control group: [{lower_con:.3f}, {upper_con:.3f}]')
print(f'ci 95% for treatment group: [{lower_treat:.3f}, {upper_treat:.3f}]')

z statistic: -0.23
p-value: 0.814
ci 95% for control group: [0.410, 0.491]
ci 95% for treatment group: [0.417, 0.498]


## Conclusion

- Since our $p$-value=0.814 is far above our $\alpha$=0.05, we cannot reject the null hypothesis 𝐻ₒ, which means that the new advertisement design did not performed better than the old one.

- There were enough data points to make a reasonable judgement.

- Finally, based on the A/B Testing Analysis the new advertisement design does not give an increase in brand awareness.

## References

- https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.sem.html

- https://www.statsmodels.org/dev/generated/statsmodels.stats.proportion.proportions_ztest.html

- https://www.yourdatateacher.com/2022/10/17/a-beginners-guide-to-statistical-hypothesis-tests/